src/borderarray.jl

```diff
@@ -0,0 +1,170 @@
+using Base: @propagate_inbounds
+
+function compatible_dimensions(arr::AbstractArray, border::Inner)
+    true
+end
+
+function compatible_dimensions(arr,border)
+    ndims(arr) == ndims(border)
+end
+
+"""
+    BorderArray(inner::AbstractArray, border::AbstractBorder) <: AbstractArray
+
+Construct a thin wrapper around the array `inner`, with given `border`. No data is copied in the constructor, instead border values are computed on the fly in `getindex` calls. Usful for stencil computations. See also [padarray](@ref).
+
+# Examples
+```julia
+julia> using ImageFiltering
+
+julia> arr = reshape(1:6, (2,3))
+2×3 reshape(::UnitRange{Int64}, 2, 3) with eltype Int64:
+ 1  3  5
+ 2  4  6
+
+julia> BorderArray(arr, Pad((1,1)))
+BorderArray{Int64,2,Base.ReshapedArray{Int64,2,UnitRange{Int64},Tuple{}},Pad{2}} with indices 0:3×0:4:
+ 1  1  3  5  5
+ 1  1  3  5  5
+ 2  2  4  6  6
+ 2  2  4  6  6
+
+julia> BorderArray(arr, Fill(10, (2,1)))
+BorderArray{Int64,2,Base.ReshapedArray{Int64,2,UnitRange{Int64},Tuple{}},Fill{Int64,2}} with indices -1:4×0:4:
+ 10  10  10  10  10
+ 10  10  10  10  10
+ 10   1   3   5  10
+ 10   2   4   6  10
+ 10  10  10  10  10
+ 10  10  10  10  10
+```
+"""
+struct BorderArray{T,N,A,B} <: AbstractArray{T,N}
+    inner::A
+    border::B
+    function BorderArray(arr::AbstractArray{T,N}, border::AbstractBorder) where {T,N}
+        if !compatible_dimensions(arr, border)
+            throw(ArgumentError("$border lacks the proper padding sizes for an array with $(ndims(arr)) dimensions"))
+        end
+        A = typeof(arr)
+        B = typeof(border)
+        new{T,N,A,B}(arr, border)
+    end
+end
+
+# these are adaped from OffsetArrays
+function Base.similar(A::BorderArray, ::Type{T}, dims::Dims) where T
+    B = similar(A.inner, T, dims)
+end
+const OffsetAxis = Union{Integer, UnitRange, Base.OneTo, IdentityUnitRange}
+function Base.similar(A::BorderArray, ::Type{T}, inds::Tuple{OffsetAxis,Vararg{OffsetAxis}}) where T
+    similar(A.inner, T, axes(A))
+end
+
+@inline function Base.axes(o::BorderArray)
+    _outeraxes(o.inner, o.border)
+end
+
+@inline function Base.size(o::BorderArray)
+    map(length, axes(o))
+end
+
+@inline function _outeraxes(arr, border::Inner)
+    axes(arr)
+end
+
+@inline function _outeraxes(arr, border)
+    map(axes(arr), border.lo, border.hi) do r, lo, hi
+        (first(r)-lo):(last(r)+hi)
+    end
+end
+
+@inline @propagate_inbounds function Base.getindex(o::BorderArray{T,N}, inds::Vararg{Int,N}) where {T,N}
+    ci = CartesianIndex(inds)
+    o[ci]
+end
+
+@inline function Base.getindex(o::BorderArray, ci::CartesianIndex)
+    if checkbounds(Bool, o.inner, ci)
+        @inbounds o.inner[ci]
+    else
+        @boundscheck checkbounds(o, ci)
+        getindex_outer_inbounds(o.inner, o.border, ci)
```

Comment:
Do we need to make sure that both paths return values of type `T`? AFAICT it's possible to do something like
```julia
v = BorderArray([1, 2], Fill(NaN, (1,)))
```
and then indexing `v` is non-inferrable.

Reply:
Good point! Do you prefer to throw an error in the constructor or a `convert` in `getindex` if that happens?

Reply:
If it's straightforward to do it in the constructor, that would be best.